web3_proxy/src/frontend/users/payment.rs

```diff
@@ -413,60 +421,26 @@ pub async fn user_balance_post(
 
         // Get the premium user-tier
         let premium_user_tier = user_tier::Entity::find()
-            .filter(user_tier::Column::Title.eq("Premium"))
-            .one(&db_conn)
+            .filter(user_tier::Column::Title.eq(UserTier::Premium.to_string()))
+            .one(&txn)
             .await?
             .context("Could not find 'Premium' Tier in user-database")?;
 
-        let txn = db_conn.begin().await?;
-        match user_balance {
-            Some(user_balance) => {
-                let balance_plus_amount = user_balance.available_balance + amount;
-                info!("New user balance is: {:?}", balance_plus_amount);
-                // Update the entry, adding the balance
-                let mut active_user_balance = user_balance.into_active_model();
-                active_user_balance.available_balance = sea_orm::Set(balance_plus_amount);
-
-                if balance_plus_amount >= Decimal::new(10, 0) {
-                    // Also make the user premium at this point ...
-                    let mut active_recipient = recipient.clone().into_active_model();
-                    // Make the recipient premium "Effectively Unlimited"
-                    active_recipient.user_tier_id = sea_orm::Set(premium_user_tier.id);
-                    active_recipient.save(&txn).await?;
-                }
-
-                debug!("New user balance model is: {:?}", active_user_balance);
-                active_user_balance.save(&txn).await?;
-                // txn.commit().await?;
-                // user_balance
-            }
-            None => {
-                // Create the entry with the respective balance
-                let active_user_balance = balance::ActiveModel {
-                    available_balance: sea_orm::ActiveValue::Set(amount),
-                    user_id: sea_orm::ActiveValue::Set(recipient.id),
-                    ..Default::default()
-                };
-
-                if amount >= Decimal::new(10, 0) {
-                    // Also make the user premium at this point ...
-                    let mut active_recipient = recipient.clone().into_active_model();
-                    // Make the recipient premium "Effectively Unlimited"
-                    active_recipient.user_tier_id = sea_orm::Set(premium_user_tier.id);
-                    active_recipient.save(&txn).await?;
-                }
-
-                info!("New user balance model is: {:?}", active_user_balance);
-                active_user_balance.save(&txn).await?;
-                // txn.commit().await?;
-                // user_balance // .try_into_model().unwrap()
-            }
-        };
+        // Gotta add to the balance entry
+        // Should be atomic, because we do a write lock during the transaction
+        let mut active_user_balance = user_balance.clone().into_active_model();
+        active_user_balance.available_balance =
+            sea_orm::Set(user_balance.available_balance + amount);
```

Comment:
Because of how referral locking is happening, I don't think that "select for update" will work well anymore.

Reply:
I removed the locking for users, only kept it for the referral table. i will test it on the referral table